feat(operator): Voting power snapshot

TimmyExogenous opened the pull request:
## Description

This PR implements the voting power snapshot mechanism.

* Adds `VotingPowerSnapshot` and `SnapshotHelper` to store the snapshots.
* Updates the snapshot at the end of the epoch or when a slash is executed.
* Updates the voting power and marks it for Dogfood when a slash is executed.
* Completes the slash logic for other AVSs.
* Prune snapshots based on the unbonding duration set by AVS.
* Add RPC and CLI support for the snapshot.
* Add the voting power snapshot initialization for Genesis AVS.
* Add unit tests
* Fix the issue regarding incorrect filter height when slashing from the pending undelegations.

### TODO
* integration tests.


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
	- Introduced new methods for querying voting power snapshots, including `QuerySnapshotHelper`, `QueryAllSnapshot`, and `QuerySpecifiedSnapshot`.
	- Added functionality to manage operator voting power and snapshots with new message types in the protocol.
	- Enhanced the test suite for AVS and operator functionalities, covering various scenarios and edge cases.

- **Bug Fixes**
	- Improved error handling across multiple methods to ensure consistent behavior when encountering zero values or invalid inputs.

- **Documentation**
	- Updated comments and documentation throughout the codebase for clarity, particularly regarding the new functionalities and error handling improvements.

- **Tests**
	- Expanded test coverage for voting power snapshots and AVS functionalities, ensuring robustness and reliability in various scenarios.

- **Chores**
	- Standardized the representation of zero values across the codebase, replacing instances of `sdkmath.NewInt(0)` and `sdkmath.LegacyNewDec(0)` with `sdkmath.ZeroInt()` and `sdkmath.LegacyZeroDec()`.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

TimmyExogenous committed:
detail the return error for operator info

TimmyExogenous committed:
Merge remote-tracking branch 'upstream/develop' into develop

TimmyExogenous committed:
Merge remote-tracking branch 'upstream/develop' into develop

TimmyExogenous committed:
Merge remote-tracking branch 'upstream/develop' into develop

TimmyExogenous committed:
Merge remote-tracking branch 'upstream/develop' into develop

TimmyExogenous committed:
Merge remote-tracking branch 'upstream/develop' into develop

TimmyExogenous committed:
Merge remote-tracking branch 'upstream/develop' into develop

TimmyExogenous committed:
Merge remote-tracking branch 'upstream/develop' into develop

TimmyExogenous committed:
Merge remote-tracking branch 'upstream/develop' into develop

TimmyExogenous committed:
Merge remote-tracking branch 'upstream/develop' into develop

TimmyExogenous committed:
Merge remote-tracking branch 'upstream/develop' into develop

TimmyExogenous committed:
Merge remote-tracking branch 'upstream/develop' into develop

TimmyExogenous committed:
Merge remote-tracking branch 'upstream/develop' into develop

TimmyExogenous committed:
Merge remote-tracking branch 'upstream/develop' into develop

TimmyExogenous committed:
Merge remote-tracking branch 'upstream/develop' into develop

TimmyExogenous committed:
Merge remote-tracking branch 'upstream/develop' into develop

TimmyExogenous committed:
Merge remote-tracking branch 'upstream/develop' into develop

TimmyExogenous committed:
Merge remote-tracking branch 'upstream/develop' into develop

TimmyExogenous committed:
Merge remote-tracking branch 'upstream/develop' into develop

TimmyExogenous committed:
Merge remote-tracking branch 'upstream/develop' into develop

TimmyExogenous committed:
add a file for validator snapshot

TimmyExogenous committed:
Merge remote-tracking branch 'upstream/develop' into develop

TimmyExogenous committed:
add the voting power snapshot data structure

TimmyExogenous committed:
Merge remote-tracking branch 'upstream/develop' into develop

TimmyExogenous committed:
implement the voting power snapshot

TimmyExogenous committed:
Merge remote-tracking branch 'upstream/develop' into develop

TimmyExogenous committed:
1. fix the issue of using incorrect height when slashing pending undelegations.
2. add the voting power pruning

TimmyExogenous committed:
fix the voting power snapshot issues identified by the unit test
add unit test for voting power snapshot

MaxMustermann2 commented on proto/exocore/operator/v1/tx.proto:
```diff
@@ -51,6 +52,71 @@ message OperatorOptedUSDValue {
+  // epoch_number indicates which epoch this snapshot serve for
+  int64  epoch_number = 5;
+  // block_time records the block time when it is stored, and it is used for snapshot pruning
+  // The epoch_number would be sufficient for pruning if we don't support AVSs changing their
```
CC @trestinlsd and @mikebraver - do we allow an AVS to change its epoch? If yes, it should be disallowed. What do you think?

TimmyExogenous replied:
right, we can remove the `block_time` field if we don't allow an AVS to change its epoch. But I am not sure if it's acceptable for an AVS. I’m not sure if there are cases where AVS is misconfigured with the wrong epoch or where the current epoch frequency configuration doesn’t meet the requirements and needs to be changed.

mikebraver replied:
Actually it'll be easier for us to handle to use unified epoch and don't support AVS to change it

TimmyExogenous replied:
I will remove the `block_time` and use the epoch number to address the pruning.

MaxMustermann2 commented on proto/exocore/operator/v1/tx.proto:
```diff
@@ -51,6 +52,71 @@ message OperatorOptedUSDValue {
+  bool has_opt_out = 2;
+
+  // has_slash is used to indicate whether there is a slash event has been executed before the epoch expires.
+  // When a slash is executed, Dogfood can immediately update the voting power based on this flag.
```
Why only dogfood and not other AVSs?

TimmyExogenous replied:
It has been fixed in the latest commit.

MaxMustermann2 commented on x/avs/keeper/keeper.go:
```diff
@@ -386,6 +388,22 @@ func (k Keeper) IterateAVSInfo(ctx sdk.Context, fn func(index int64, avsInfo typ
+func (k Keeper) GetAVSEpochInfo(ctx sdk.Context, addr string) (*epochstypes.EpochInfo, error) {
+	store := prefix.NewStore(ctx.KVStore(k.storeKey), types.KeyPrefixAVSInfo)
+	value := store.Get(common.HexToAddress(addr).Bytes())
+	if value == nil {
+		return nil, types.ErrNoKeyInTheStore.Wrapf("GetAVSInfo: key is %s", addr)
+	}
```
+1

TimmyExogenous replied:
Thanks, it will be fixed in the next commit.

MaxMustermann2 commented on x/avs/keeper/keeper.go:
```diff
@@ -386,6 +388,22 @@
 	}
 }
 
+func (k Keeper) GetAVSEpochInfo(ctx sdk.Context, addr string) (*epochstypes.EpochInfo, error) {
+	store := prefix.NewStore(ctx.KVStore(k.storeKey), types.KeyPrefixAVSInfo)
+	value := store.Get(common.HexToAddress(addr).Bytes())
+	if value == nil {
+		return nil, types.ErrNoKeyInTheStore.Wrapf("GetAVSInfo: key is %s", addr)
+	}
+	avsInfo := types.AVSInfo{}
+	k.cdc.MustUnmarshal(value, &avsInfo)
+
+	epochInfo, found := k.epochsKeeper.GetEpochInfo(ctx, avsInfo.EpochIdentifier)
+	if !found {
+		return nil, types.ErrEpochNotFound.Wrapf("epoch info not found %s", avsInfo.EpochIdentifier)
+	}
+	return &epochInfo, nil
+}
```
+1

TimmyExogenous replied:
Thanks, it will be fixed in the next commit.

MaxMustermann2 commented on x/dogfood/keeper/abci.go:
```diff
@@ -15,12 +15,16 @@ func (k Keeper) BeginBlock(ctx sdk.Context) {
 
 func (k Keeper) EndBlock(ctx sdk.Context) []abci.ValidatorUpdate {
-	if !k.IsEpochEnd(ctx) {
+	chainIDWithoutRevision := avstypes.ChainIDWithoutRevision(ctx.ChainID())
```
Instead of doing it this way, may I propose using slashing hooks? Whenever an operator is slashed in x/operator, it calls all of the registered hooks implemented in other modules (similar to `impl_epochs_hooks.go`, `impl_operator_hooks.go` and `impl_delegation_hooks.go`), which can then decide what to do in response to the slashing.

TimmyExogenous replied:
It seems the hook can't trigger an `abci.ValidatorUpdate` for CometBFT, so we still need to set a flag to indicate whether to update the validator in EndBlock.

MaxMustermann2 replied:
While that is true, the hook implementation can mark within itself `k.MarkEpochEnd(ctx)` and then processing of the new validator set will proceed as usual at the end of the block. This approach will help decouple the data structure requirements between x/operator and x/dogfood or x/avs. In other words, this will allow dropping of the `has_slash` flag from the `SnapshotHelper`.

To avoid confusion, `MarkEpochEnd` could be renamed to something like `ShouldUpdateValidatorSet`, and the flag will be set upon epoch end and in the slashing hook implementation.

TimmyExogenous replied:
right, I will change it in the next commit.

MaxMustermann2 commented on x/operator/keeper/abci.go:
```diff
@@ -10,25 +13,35 @@
-		ctx.Logger().Info("UpdateVotingPower the assets list supported by AVS is nil or can't get the assets list", "error", err)
+	if getAssetsErr != nil || assets == nil || getSelfDelegationErr != nil {
+		ctx.Logger().Info("UpdateVotingPower the assets list supported by AVS is nil or can't get AVS info", "getAssetsErr", getAssetsErr, "getSelfDelegationErr", getSelfDelegationErr)
+		// using cache context to ensure the atomicity of the operation.
```
+1

TimmyExogenous replied:
Thanks, it will be fixed in the next commit.

MaxMustermann2 commented on x/operator/keeper/abci.go:
```diff
@@ -90,10 +107,90 @@ func (k *Keeper) UpdateVotingPower(ctx sdk.Context, avsAddr string) error {
+	// When the snapshot helper does not exist, it represents the initial state of AVS,
+	// where no snapshot information has been stored. Therefore, it is necessary to store
+	// both the snapshot and the helper information.
+	var snapshotHelper operatortypes.SnapshotHelper
```
```suggestion
	snapshotHelper := operatortypes.SnapshotHelper{}
```

TimmyExogenous replied:
Thanks, it will be updated in the next commit.

MaxMustermann2 commented on x/operator/keeper/abci.go:
```diff
@@ -90,10 +107,90 @@ func (k *Keeper) UpdateVotingPower(ctx sdk.Context, avsAddr string) error {
+		return operatortypes.ErrFailToClearVPSnapshot.Wrapf("ClearVotingPowerSnapshot: failed to get the avs epoch information, err:%s, avs:%s", err, avs)
+	}
+
+	clearTime := ctx.BlockTime().Add(-epochInfo.Duration * time.Duration(unbondingDuration))
```
I don't think that this is always correct because of the lag in sending / receiving IBC messages. Maybe a callback can be created (to be called by the AVS) to clear it when the AVS knows that a snapshot is not being used.

TimmyExogenous replied:
It can be applied to AVSs that don’t need to address lag, such as Dogfood, and AVSs where all system logic is deployed on the Exocore chain. For appChain-type AVSs, we might consider using time or epoch numbers—obtainable from the callback—to replace `ctx.BlockTime()`. However, if we allow AVSs to change their epoch configuration, retrieving the block time may become challenging. 

We can adjust this part of the implementation after merging the app chain PR.

MaxMustermann2 commented on x/operator/keeper/voting_power_snapshot.go:
```diff
@@ -0,0 +1,130 @@
+		for ; iterator.Valid(); iterator.Next() {
+			var amounts assetstype.OperatorAssetInfo
+			k.cdc.MustUnmarshal(iterator.Value(), &amounts)
+			if string(iterator.Key()) <= string(keyWithHeight) {
```
Instead of using a string, use `sdk.BigEndianToUint64` or similar

TimmyExogenous replied:
Yes, it has been fixed in the latest commit.

MaxMustermann2 commented on x/operator/keeper/abci.go:
```diff
@@ -90,10 +107,90 @@ func (k *Keeper) UpdateVotingPower(ctx sdk.Context, avsAddr string) error {
+	if err != nil {
+		return err
+	}
+	snapshotKey := assetstype.GetJoinedStoreKey(avsAddr, strconv.FormatInt(snapshotHeight, 10))
```
This is not sortable in the iterator. Please use the bytes format as `sdk.Uint64ToBigEndian(height)`

TimmyExogenous replied:
Yes, it has been fixed in the latest commit.

TimmyExogenous commented on x/operator/keeper/abci.go:
```diff
@@ -90,6 +111,63 @@
+	// where no snapshot information has been stored. Therefore, it is necessary to store
+	// both the snapshot and the helper information.
+	var snapshotHelper operatortypes.SnapshotHelper
+	if !k.HasSnapshotHelper(cc, avsAddr) {
+		isSnapshotChanged = true
+	} else {
+		snapshotHelper, err = k.GetSnapshotHelper(cc, avsAddr)
+		if err != nil {
+			return err
+		}
+	}
+	votingPowerSnapshot := operatortypes.VotingPowerSnapshot{
```
It will be fixed in the next commit.

TimmyExogenous commented on proto/exocore/operator/v1/tx.proto:
```diff
@@ -51,6 +51,62 @@ message OperatorOptedUSDValue {
+  ];
+  // operator_voting_powers records the active voting power of all operators
+  // for the specified AVS
+  repeated OperatorVotingPower voting_power_set = 3;
```
Thanks, it will be fixed in the next commit.

TimmyExogenous commented on x/operator/keeper/abci.go:
```diff
@@ -90,10 +107,90 @@ func (k *Keeper) UpdateVotingPower(ctx sdk.Context, avsAddr string) error {
+	if err != nil {
+		return err
+	}
+	snapshotKey := assetstype.GetJoinedStoreKey(avsAddr, strconv.FormatInt(snapshotHeight, 10))
+	err = k.SetVotingPowerSnapshot(cc, snapshotKey, &votingPowerSnapshot)
+	if err != nil {
+		return err
+	}
+
 	writeFunc()
 	return nil
 }
 
+func (k *Keeper) ClearVotingPowerSnapshot(ctx sdk.Context, avs string) error {
+	// calculate the time before which the snapshot should be cleared.
+	unbondingDuration, err := k.avsKeeper.GetAVSUnbondingDuration(ctx, avs)
+	if err != nil {
+		return operatortypes.ErrFailToClearVPSnapshot.Wrapf("ClearVotingPowerSnapshot: failed to get the avs unbonding duration, err:%s, avs:%s", err, avs)
+	}
+	epochInfo, err := k.avsKeeper.GetAVSEpochInfo(ctx, avs)
+	if err != nil {
+		return operatortypes.ErrFailToClearVPSnapshot.Wrapf("ClearVotingPowerSnapshot: failed to get the avs epoch information, err:%s, avs:%s", err, avs)
+	}
+
+	clearTime := ctx.BlockTime().Add(-epochInfo.Duration * time.Duration(unbondingDuration))
+	err = k.RemoveVotingPowerSnapshot(ctx, avs, clearTime)
+	if err != nil {
+		ctx.Logger().Error("Failed to get the avs epoch information", "avs", avs, "error", err)
+		return operatortypes.ErrFailToClearVPSnapshot.Wrapf("ClearVotingPowerSnapshot: failed to remove voting power snapshot, err:%s, avs:%s", err, avs)
+	}
+	return nil
+}
```
Thanks, it will be fixed in the next commit.

TimmyExogenous commented on x/operator/keeper/slash.go:
```diff
@@ -156,6 +154,22 @@ func (k *Keeper) Slash(ctx sdk.Context, parameter *types.SlashInputInfo) error {
+		}
+		parameter.Power = votingPower.VotingPower.TruncateInt64()
+		if parameter.Power < 0 {
+			return types.ErrInvalidSlashPower.Wrapf("slash: valid voting power, the power is:%v", parameter.Power)
```
It will be fixed in the next commit.

TimmyExogenous commented on x/operator/keeper/impl_epoch_hook.go:
```diff
@@ -30,12 +30,19 @@ func (wrapper EpochsHooksWrapper) AfterEpochEnd(
 	for _, avs := range avsList {
 		// avs address should be hex
-		err := wrapper.keeper.UpdateVotingPower(ctx, avs)
+		err := wrapper.keeper.UpdateVotingPower(ctx, avs, epochIdentifier, epochNumber, false)
 		if err != nil {
 			ctx.Logger().Error("Failed to update voting power", "avs", avs, "error", err)
 			// Handle the error gracefully, continue to the next AVS
 			continue
 		}
```
Currently, it's unnecessary.

TimmyExogenous commented on x/operator/keeper/impl_epoch_hook.go:
```diff
@@ -30,12 +30,19 @@
 	for _, avs := range avsList {
 		// avs address should be hex
-		err := wrapper.keeper.UpdateVotingPower(ctx, avs)
+		err := wrapper.keeper.UpdateVotingPower(ctx, avs, epochIdentifier, epochNumber, false)
 		if err != nil {
 			ctx.Logger().Error("Failed to update voting power", "avs", avs, "error", err)
 			// Handle the error gracefully, continue to the next AVS
 			continue
 		}
+		// clear the expired voting power snapshot.
+		err = wrapper.keeper.ClearVotingPowerSnapshot(ctx, avs)
+		if err != nil {
+			ctx.Logger().Error("Failed to clear voting power snapshot", "avs", avs, "error", err)
+			// Handle the error gracefully, continue to the next AVS
+			continue
+		}
```
Currently, it's unnecessary because there won't be a significant impact even if the snapshot isn't cleared when the voting power update fails.

TimmyExogenous commented on x/operator/keeper/setup_test.go:
```diff
@@ -18,16 +18,14 @@ type OperatorTestSuite struct {
-	depositAmount         sdkmath.Int
-	delegationAmount      sdkmath.Int
-	updatedAmountForOptIn sdkmath.Int
+	operatorAddr     sdk.AccAddress
+	avsAddr          string
+	assetID          string
+	stakerID         string
+	assetAddr        common.Address
+	assetDecimal     uint32
+	clientChainLzID  uint64
+	delegationAmount sdkmath.Int
```
This refactor isn't too necessary.

TimmyExogenous commented on x/operator/keeper/abci.go:
```diff
@@ -90,10 +110,97 @@ func (k *Keeper) UpdateVotingPower(ctx sdk.Context, avsAddr string) error {
+	clearTime := ctx.BlockTime().Add(-epochInfo.Duration * time.Duration(unbondingDuration))
+	err = k.RemoveVotingPowerSnapshot(ctx, avs, clearTime)
+	if err != nil {
+		ctx.Logger().Error("Failed to get the avs epoch information", "avs", avs, "error", err)
+		return operatortypes.ErrFailToClearVPSnapshot.Wrapf("ClearVotingPowerSnapshot: failed to remove voting power snapshot, err:%s, avs:%s", err, avs)
```
It will be fixed in the next commit.

TimmyExogenous commented on x/operator/keeper/voting_power_snapshot.go:
```diff
@@ -0,0 +1,207 @@
+	// If there is no snapshot for the input height, we need to find the correct key.
+	// The snapshot closest to the input height is the one used for its voting
+	// power information. The correct snapshot key can be found by taking advantage
+	// of the ascending order of data returned when using an iterator range.
+	avsEthAddr := common.HexToAddress(avsAddr)
+	findKey := types.KeyForVotingPowerSnapshot(avsEthAddr, height)
+	findHeight := height
+	if !store.Has(findKey) {
+		iterator := sdk.KVStorePrefixIterator(store, avsEthAddr.Bytes())
+		defer iterator.Close()
+		for ; iterator.Valid(); iterator.Next() {
+			_, keyHeight, err := types.ParseVotingPowerSnapshotKey(iterator.Key())
+			if err != nil {
+				return 0, nil, err
+			}
+			if keyHeight <= height {
+				findKey = iterator.Key()
+				findHeight = keyHeight
+			} else {
+				break
+			}
+		}
+	}
+
+	value := store.Get(findKey)
+	if value == nil {
+		return 0, nil, types.ErrNoKeyInTheStore.Wrapf("LoadVotingPowerSnapshot: height is %v", findHeight)
+	}
+	k.cdc.MustUnmarshal(value, &ret)
+
+	// fall back to get the snapshot if the key height doesn't equal to the `LastChangedHeight`
+	if findHeight != ret.LastChangedHeight {
+		value = store.Get(types.KeyForVotingPowerSnapshot(common.HexToAddress(avsAddr), ret.LastChangedHeight))
+		if value == nil {
+			return 0, nil, types.ErrNoKeyInTheStore.Wrapf("LoadVotingPowerSnapshot: fall back to the height %v", ret.LastChangedHeight)
+		}
+		k.cdc.MustUnmarshal(value, &ret)
+	}
+	return findHeight, &ret, nil
+}
```
The voting power set will be null if the `findHeight` isn't equal to `ret.LastChangedHeight`, so it's expected to fallback to the correct key.

TimmyExogenous commented on x/operator/keeper/voting_power_snapshot.go:
```diff
@@ -0,0 +1,207 @@
+func (k *Keeper) SetOptOutFlag(ctx sdk.Context, avsAddr string, hasOptOut bool) error {
+	opFunc := func(helper *types.SnapshotHelper) error {
+		helper.HasOptOut = hasOptOut
+		return nil
```
I will add this code comment in the next commit.

TimmyExogenous commented on x/operator/keeper/voting_power_snapshot.go:
```diff
@@ -0,0 +1,207 @@
+	return nil
+}
+
+func (k *Keeper) UpdateSnapshotHelper(ctx sdk.Context, avsAddr string, opFunc func(helper *types.SnapshotHelper) error) error {
+	store := prefix.NewStore(ctx.KVStore(k.storeKey), types.KeyPrefixSnapshotHelper)
+	var snapshotHelper types.SnapshotHelper
+	value := store.Get([]byte(avsAddr))
+	if value != nil {
+		k.cdc.MustUnmarshal(value, &snapshotHelper)
+	}
+	err := opFunc(&snapshotHelper)
+	if err != nil {
+		return err
+	}
+	bz := k.cdc.MustMarshal(&snapshotHelper)
+	store.Set([]byte(avsAddr), bz)
+	return nil
+}
```
It won't be called concurrently, so we don't consider this issue now.

TimmyExogenous commented on x/operator/keeper/voting_power_snapshot.go:
```diff
@@ -0,0 +1,207 @@
+	return store.Has([]byte(avsAddr))
+}
+
+func (k *Keeper) HasSlash(ctx sdk.Context, avsAddr string) bool {
+	helper, err := k.GetSnapshotHelper(ctx, avsAddr)
+	if err != nil {
+		return false
+	}
+	return helper.HasSlash
+}
```
I will change it in the next commit.

TimmyExogenous commented on x/operator/keeper/voting_power_snapshot.go:
```diff
@@ -0,0 +1,207 @@
+	var ret types.VotingPowerSnapshot
+	value := store.Get(key)
+	if value == nil {
+		return nil, types.ErrNoKeyInTheStore.Wrapf("GetVotingPowerSnapshot: key is %s", key)
```
thanks, it will be fixed in the next commit.

TimmyExogenous commented on x/operator/keeper/voting_power_snapshot_test.go:
```diff
@@ -0,0 +1,303 @@
+		err = suite.App.OperatorKeeper.OptIn(suite.Ctx, operators[i], suite.avsAddr)
+		suite.NoError(err)
+	}
+	sort.Slice(stakers, func(i, j int) bool {
+		return operators[i].String() < operators[j].String()
+	})
```
It's a requirement for one-to-one correspondence, so it's expected.

TimmyExogenous committed:
fix the issues identified by code comment

bwhour commented on x/operator/types/utils.go:
```diff
@@ -25,3 +25,13 @@ func ChainIDWithLenKey(chainID string) []byte {
 }
+
+func GetSpecifiedVotingPower(operator string, votingPowerSet []*OperatorVotingPower) *OperatorVotingPower {
+	for idx := range votingPowerSet {
```
Why not use pointers to traverse and avoid copying? 

> for _, vp := range votingPowers {
        if vp.OperatorAddr == operator {
            return vp
        }
    }

TimmyExogenous replied:
thanks, it will be changed in the next commit.

TimmyExogenous committed:
1. add cli and rpc for voting power snapshot
2. use a hook to notify the Dogfood slash event
3. initialize the voting power snapshot for genesis AVSs
4. add a function to get the impactful AVS list for an operator

TimmyExogenous committed:
fix the issues regarding code comments
fix the lint error

TimmyExogenous commented on x/operator/keeper/voting_power_snapshot_test.go:
```diff
@@ -0,0 +1,303 @@
+	suite.NoError(err)
+	suite.True(snapshotHelper.HasSlash)
+	suite.Equal(suite.Ctx.BlockHeight(), snapshotHelper.LastChangedHeight)
+}
```
It's a test for the voting power snapshot. There's no difference whether the slashing is applied to Dogfood or other AVS.

TimmyExogenous commented on x/operator/keeper/voting_power_snapshot_test.go:
```diff
@@ -0,0 +1,303 @@
+	suite.Equal(suite.Ctx.BlockHeight(), snapshotAfterSlash.LastChangedHeight)
+	operatorVPBeforeSlash := types.GetSpecifiedVotingPower(testHelper.operators[index].String(), snapshotBeforeSlash.OperatorVotingPowers)
+	operatorVPAfterSlash := types.GetSpecifiedVotingPower(testHelper.operators[index].String(), snapshotAfterSlash.OperatorVotingPowers)
+	suite.Equal(operatorVPBeforeSlash.VotingPower.Mul(remainingProportion), operatorVPAfterSlash.VotingPower)
+
+	snapshotHelper, err := suite.App.OperatorKeeper.GetSnapshotHelper(suite.Ctx, suite.avsAddr)
+	suite.NoError(err)
+	suite.True(snapshotHelper.HasSlash)
+	suite.Equal(suite.Ctx.BlockHeight(), snapshotHelper.LastChangedHeight)
+}
+
+func (suite *OperatorTestSuite) TestSnapshotPruning() {
+	testHelper := suite.prepareForSnapshotTesting(operatorNumber)
+	suite.runToEpochEnd()
+	firstSnapshotHeight := suite.Ctx.BlockHeight()
+
+	avsUnbondingDuration, err := suite.App.AVSManagerKeeper.GetAVSUnbondingDuration(suite.Ctx, suite.avsAddr)
+	suite.NoError(err)
+
+	runEpochNumber := avsUnbondingDuration + 2
+	for i := uint64(0); i < runEpochNumber; i++ {
+		suite.runToEpochEnd()
+	}
+	_, _, err = suite.App.OperatorKeeper.LoadVotingPowerSnapshot(suite.Ctx, suite.avsAddr, firstSnapshotHeight)
+	suite.NoError(err)
+
+	key := types.KeyForVotingPowerSnapshot(common.HexToAddress(suite.avsAddr), firstSnapshotHeight+blockNumberPerEpoch)
+	_, err = suite.App.OperatorKeeper.GetVotingPowerSnapshot(suite.Ctx, key)
+	suite.Error(err)
+
+	// change the voting power of the operator at index 0.
+	index := 0
+	suite.prepareDelegation(true, testHelper.stakers[index], usdtAddr, testHelper.operators[index], testHelper.delegateAmount)
+	for i := uint64(0); i < runEpochNumber; i++ {
+		suite.runToEpochEnd()
+	}
+	key = types.KeyForVotingPowerSnapshot(common.HexToAddress(suite.avsAddr), firstSnapshotHeight)
+	_, err = suite.App.OperatorKeeper.GetVotingPowerSnapshot(suite.Ctx, key)
+	suite.Error(err)
+}
```
Thanks, it will be added in the next commit.

TimmyExogenous commented on x/dogfood/keeper/impl_operator_hooks.go:
```diff
@@ -91,3 +91,29 @@ func (h OperatorHooksWrapper) AfterOperatorKeyRemovalInitiated(
 	}
 }
+
+func (h OperatorHooksWrapper) AfterSlash(
+	ctx sdk.Context, operator sdk.AccAddress, affectedAVSList []string,
+) {
+	chainIDWithoutRevision := avstypes.ChainIDWithoutRevision(ctx.ChainID())
+	dogfoodAVSAddr := avstypes.GenerateAVSAddr(chainIDWithoutRevision)
+	for i := range affectedAVSList {
+		if affectedAVSList[i] == dogfoodAVSAddr {
+			// check if the operator is in the current validator set.
+			found, wrappedKey, err := h.keeper.operatorKeeper.GetOperatorConsKeyForChainID(ctx, operator, chainIDWithoutRevision)
+			if !found || err != nil {
+				ctx.Logger().Error("AfterSlash the consensus key isn't found by the chainIDWithoutRevision and operator address", "operatorAddr", operator, "chainIDWithoutRevision", chainIDWithoutRevision, "err", err)
+				return
+			}
+			// check if the key is active yet
+			isValidator := false
+			_, isValidator = h.keeper.GetExocoreValidator(
+				ctx, wrappedKey.ToConsAddr(),
+			)
+			if isValidator {
+				h.keeper.MarkUpdateValidatorSetFlag(ctx)
+			}
+			break
+		}
+	}
+}
```
It's unnecessary; the DoS should be addressed through gas costs and permission controls for the registration and opting-in of AVS and operator.

TimmyExogenous commented on x/operator/keeper/opt_test.go:
```diff
@@ -84,20 +86,21 @@ func (suite *OperatorTestSuite) prepareDelegation(isDelegation bool, assetAddr c
 
-func (suite *OperatorTestSuite) prepareAvs(assetIDs []string) {
+func (suite *OperatorTestSuite) prepareAvs(assetIDs []string, epochIdentifier string) {
+	suite.avsAddr = common.BytesToAddress([]byte("avsTestAddr")).String()
 	err := suite.App.AVSManagerKeeper.UpdateAVSInfo(suite.Ctx, &avstypes.AVSRegisterOrDeregisterParams{
 		Action:          avskeeper.RegisterAction,
-		EpochIdentifier: epochstypes.HourEpochID,
+		EpochIdentifier: epochIdentifier,
 		AvsAddress:      suite.avsAddr,
 		AssetID:         assetIDs,
+		UnbondingPeriod: 5,
 	})
 	suite.NoError(err)
```
There are different requirements in the other unit tests.

TimmyExogenous commented on x/operator/keeper/operator.go:
```diff
@@ -207,37 +271,38 @@ func (k *Keeper) SetAllOptedInfo(ctx sdk.Context, optedStates []operatortypes.Op
-		}
-		if avsAddr == keys[1] {
-			operatorList = append(operatorList, keys[0])
+	opFunc := func(key []byte, optedInfo *operatortypes.OptedInfo) error {
+		if optedInfo.OptedOutHeight == operatortypes.DefaultOptedOutHeight {
+			keys, err := assetstype.ParseJoinedStoreKey(key, 2)
+			if err != nil {
+				return err
+			}
+			if avsAddr == keys[1] {
+				operatorList = append(operatorList, keys[0])
+			}
 		}
+		return nil
+	}
+	err := k.IterateOptInfo(ctx, false, []byte{}, opFunc)
+	if err != nil {
+		return nil, err
```
Since there won’t be many AVSs or operators, we’re not currently considering this suggestion.

TimmyExogenous commented on x/operator/keeper/operator.go:
```diff
@@ -179,19 +179,83 @@
+}
+
+func (k *Keeper) GetOptedInAVSForOperator(ctx sdk.Context, operatorAddr string) ([]string, error) {
+	avsList := make([]string, 0)
+	opFunc := func(key []byte, optedInfo *operatortypes.OptedInfo) error {
+		if optedInfo.OptedOutHeight == operatortypes.DefaultOptedOutHeight {
+			keys, err := assetstype.ParseJoinedStoreKey(key, 2)
+			if err != nil {
+				return err
+			}
+			avsList = append(avsList, keys[1])
 		}
-		avsList = append(avsList, keys[1])
+		return nil
+	}
+	err := k.IterateOptInfo(ctx, false, []byte(operatorAddr), opFunc)
+	if err != nil {
+		return nil, err
+	}
+	return avsList, nil
+}
```
We don’t know the size of the AVS list in advance, but memory allocation won’t be excessive since the number of AVS items is relatively small.

TimmyExogenous commented on x/operator/keeper/usd_value.go:
```diff
@@ -269,7 +270,7 @@ func (k *Keeper) SetAllOperatorUSDValues(ctx sdk.Context, usdValues []operatorty
 		usdValue := usdValues[i]
 		bz := k.cdc.MustMarshal(&usdValue.OptedUSDValue)
-		store.Set([]byte(usdValue.Key), bz)
+		store.Set([]byte(strings.ToLower(usdValue.Key)), bz)
```
It's fine to do this change, because we use lower case in the whole system.

TimmyExogenous commented on x/operator/keeper/voting_power_snapshot.go:
```diff
@@ -0,0 +1,268 @@
+}
+
+func (k *Keeper) SetSnapshotHelper(ctx sdk.Context, avsAddr string, snapshotHelper *types.SnapshotHelper) error {
+	store := prefix.NewStore(ctx.KVStore(k.storeKey), types.KeyPrefixSnapshotHelper)
+	bz := k.cdc.MustMarshal(snapshotHelper)
+	store.Set([]byte(avsAddr), bz)
+	return nil
+}
```
Thanks, it will be fixed in the next commit.

TimmyExogenous commented on x/operator/keeper/voting_power_snapshot.go:
```diff
@@ -0,0 +1,268 @@
+		epochNumber := epochInfo.CurrentEpoch
+		// set the epoch number to 1 when epoch start for the first time.
+		if !epochInfo.EpochCountingStarted {
+			epochNumber = 1
```
thanks, it will be fixed in the next commit.

TimmyExogenous commented on proto/exocore/operator/v1/query.proto:
```diff
@@ -221,6 +221,55 @@ message QueryOptInfoRequest {
   OperatorAVSAddress operator_and_avs = 1 [(gogoproto.embed) = true];
 }
 
+// QuerySnapshotAndHelperRequest is the request to obtain the voting power snapshot
+// and helper information.
+message QuerySnapshotAndHelperRequest {
+  // avs address
+  string avs = 1;
+}
```
Only one record will be returned, so it's unnecessary to use pagination.

TimmyExogenous commented on proto/exocore/operator/v1/query.proto:
```diff
@@ -221,6 +221,55 @@
+  string avs = 1;
+}
+
+// QuerySpecifiedSnapshotRequest is the request to obtain the voting power snapshot
+// at the specified height.
+message QuerySpecifiedSnapshotRequest {
+  // avs address
+  string avs = 1;
+  // height is used to specify the snapshot height you want to query.
+  int64 height = 2;
+}
```
thanks, the check will be added when retrieving states

TimmyExogenous commented on proto/exocore/operator/v1/query.proto:
```diff
@@ -313,4 +362,19 @@
+  // QuerySnapshotHelper queries the snapshot helper of the AVS
+  rpc QuerySpecifiedSnapshot(QuerySpecifiedSnapshotRequest) returns (VotingPowerSnapshotWithKeyHeight) {
+    option (google.api.http).get = "/exocore/operator/v1/snapshot/{avs}/{height}";
+  }
```
Thanks, it will be fixed in the next commit.

TimmyExogenous replied:
Thanks, it will be fixed in the next commit.

TimmyExogenous commented on proto/exocore/operator/v1/query.proto:
```diff
@@ -278,8 +327,8 @@
     option (google.api.http).get =
-      "/exocore/operator/v1/QueryOperatorSlashInfo/{operator_and_avs.operator_addr}/"
-      "{operator_and_avs.avs_address}";
+    "/exocore/operator/v1/QueryOperatorSlashInfo/{operator_and_avs.operator_addr}/"
+        "{operator_and_avs.avs_address}";
```
thanks, it will be fixed in the next commit.

TimmyExogenous commented on proto/exocore/operator/v1/query.proto:
```diff
@@ -266,8 +315,8 @@
     option (google.api.http).get =
-      "/exocore/operator/v1/QueryOperatorUSDValue/{operator_and_avs.operator_addr}/"
-      "{operator_and_avs.avs_address}";
+    "/exocore/operator/v1/QueryOperatorUSDValue/{operator_and_avs.operator_addr}/"
+        "{operator_and_avs.avs_address}";
```
thanks, it will be fixed in the next commit

TimmyExogenous commented on proto/exocore/operator/v1/query.proto:
```diff
@@ -313,4 +362,19 @@ service Query {
     };
   }
+
+  // QuerySnapshotHelper queries the snapshot helper of the AVS
+  rpc QuerySnapshotHelper(QuerySnapshotAndHelperRequest) returns (SnapshotHelper) {
+    option (google.api.http).get = "/exocore/operator/v1/snapshot_helper/{avs}";
+  }
+
+  // QuerySnapshotHelper queries the snapshot helper of the AVS
+  rpc QuerySpecifiedSnapshot(QuerySpecifiedSnapshotRequest) returns (VotingPowerSnapshotKeyHeight) {
+    option (google.api.http).get = "/exocore/operator/v1/snapshot/{avs}/{height}";
+  }
```
thanks, it will be fixed in the next commit.

TimmyExogenous commented on x/operator/keeper/grpc_query.go:
```diff
@@ -253,3 +257,51 @@ func (k *Keeper) QueryOptInfo(goCtx context.Context, req *types.QueryOptInfoRequ
 	return k.GetOptedInfo(ctx, req.OperatorAddr, req.AvsAddress)
 }
+
+func (k *Keeper) QuerySnapshotHelper(goCtx context.Context, req *types.QuerySnapshotHelperRequest) (*types.SnapshotHelper, error) {
+	ctx := sdk.UnwrapSDKContext(goCtx)
+	snapshotHelper, err := k.GetSnapshotHelper(ctx, strings.ToLower(req.Avs))
+	if err != nil {
+		return nil, err
+	}
+	return &snapshotHelper, nil
+}
```
Thanks, I will add the check when retrieving states

TimmyExogenous commented on x/operator/keeper/grpc_query.go:
```diff
@@ -253,3 +257,51 @@
+	return &snapshotHelper, nil
+}
+
+func (k *Keeper) QuerySpecifiedSnapshot(goCtx context.Context, req *types.QuerySpecifiedSnapshotRequest) (*types.VotingPowerSnapshotKeyHeight, error) {
+	ctx := sdk.UnwrapSDKContext(goCtx)
+	findHeight, snapshot, err := k.LoadVotingPowerSnapshot(ctx, strings.ToLower(req.Avs), req.Height)
+	if err != nil {
+		return nil, err
+	}
+	return &types.VotingPowerSnapshotKeyHeight{
+		SnapshotKeyHeight: findHeight,
+		Snapshot:          snapshot,
+	}, nil
+}
```
Thanks, the check will be added in `GetSnapshotHeightAndKey`

TimmyExogenous commented on x/operator/keeper/grpc_query.go:
```diff
@@ -253,3 +257,51 @@
+	}, nil
+}
+
+func (k *Keeper) QueryAllSnapshot(goCtx context.Context, req *types.QueryAllSnapshotRequest) (*types.QueryAllSnapshotResponse, error) {
+	ctx := sdk.UnwrapSDKContext(goCtx)
+	res := make([]*types.VotingPowerSnapshotKeyHeight, 0)
+
+	snapshotPrefix := types.AppendMany(types.KeyPrefixVotingPowerSnapshot, common.HexToAddress(req.Avs).Bytes())
+	store := prefix.NewStore(ctx.KVStore(k.storeKey), snapshotPrefix)
+	pageRes, err := query.Paginate(store, req.Pagination, func(key []byte, value []byte) error {
+		ret := &types.VotingPowerSnapshot{}
+		// don't use MustUnmarshal to not panic for queries
+		if err := ret.Unmarshal(value); err != nil {
+			return err
+		}
+		res = append(res, &types.VotingPowerSnapshotKeyHeight{
+			SnapshotKeyHeight: int64(binary.BigEndian.Uint64(key)), // #nosec G115
+			Snapshot:          ret,
+		})
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+	return &types.QueryAllSnapshotResponse{
+		Snapshots:  res,
+		Pagination: pageRes,
+	}, nil
+}
```
thanks, it will fixed in the next commit.

TimmyExogenous commented on x/operator/keeper/voting_power_snapshot.go:
```diff
@@ -0,0 +1,279 @@
+		}
+		return nil, types.ErrNoKeyInTheStore.Wrapf("GetVotingPowerSnapshot: invalid key, avs:%s height:%v", avsAddr, height)
+	}
+	k.cdc.MustUnmarshal(value, &ret)
+	return &ret, nil
```
It's unnecessary, it follows the Cosmos-SDK code style.

TimmyExogenous commented on x/operator/keeper/voting_power_snapshot.go:
```diff
@@ -0,0 +1,279 @@
+		defer iterator.Close()
+		for ; iterator.Valid(); iterator.Next() {
+			_, keyHeight, err := types.ParseVotingPowerSnapshotKey(iterator.Key())
+			if err != nil {
+				return 0, nil, err
+			}
+			if keyHeight <= height {
+				findKey = iterator.Key()
+				findHeight = keyHeight
+			} else {
+				break
+			}
+		}
```
The `findKey` has been initialized, it's unnecessary to check the existence here.

TimmyExogenous committed:
Merge remote-tracking branch 'upstream/develop' into develop

TimmyExogenous committed:
Merge remote-tracking branch 'upstream/develop' into develop

TimmyExogenous committed:
rebase to develop and fix the conflict

TimmyExogenous committed:
Merge remote-tracking branch 'upstream/develop' into develop

cloud8little commented on x/operator/client/cli/query.go:
```diff
@@ -456,3 +460,106 @@ func GetOptInfo() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "QueryAllSnapshot <avsAddr>",
+		Short: "Get the all voting power snapshots for the avs",
+		Long:  "Get the all voting power snapshots for the avs",
```
```suggestion
		Long:  "Get the latest 6 epochs voting power snapshots for the avs",
```

TimmyExogenous replied:
Seems like it should be the `unbondingDuration+1`. Six is for the dogfood AVS.

cloud8little commented on x/operator/client/cli/query.go:
```diff
@@ -456,3 +460,106 @@ func GetOptInfo() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "QuerySpecifiedSnapshot <avsAddr> <height>",
+		Short: "Get the AVS voting power snapshot at specified height",
+		Long:  "Get the AVS voting power snapshot at specified height",
```
```suggestion
		Long:  "Get the AVS voting power snapshot at specified height, support latest 6 epochs",
```

bwhour commented on precompiles/avs/avs_test.go:
```diff
@@ -1,13 +1,14 @@
 	"strconv"
 	"time"
 
+	"cosmossdk.io/math"
+	assetstypes "github.com/ExocoreNetwork/exocore/x/assets/types"
+	"github.com/ExocoreNetwork/exocore/x/avs/types"
+
```
@coderabbitai  they should be different library.

bwhour commented on x/operator/keeper/abci.go:
```diff
@@ -47,18 +54,18 @@ func (k *Keeper) UpdateVotingPower(ctx sdk.Context, avsAddr string) error {
 		// clear the old voting power for the operator
+		lastOptedUSDValue := *optedUSDValues
 		*optedUSDValues = operatortypes.OperatorOptedUSDValue{
 			TotalUSDValue:  sdkmath.LegacyNewDec(0),
```
Reduce unnecessary object creation: use sdkmath.**LegacyZeroDec**() instead of sdkmath.LegacyNewDec(0).

TimmyExogenous replied:
Thanks, it will be fixed in the next commit.

TimmyExogenous commented on x/operator/keeper/slash.go:
```diff
@@ -71,7 +69,7 @@ func (k *Keeper) CheckSlashParameter(ctx sdk.Context, parameter *types.SlashInpu
 // stakers should be cleared, because the divisor will be zero when calculating the share
 // of new delegation after the slash.
-func (k *Keeper) SlashAssets(ctx sdk.Context, parameter *types.SlashInputInfo) (*types.SlashExecutionInfo, error) {
+func (k *Keeper) SlashAssets(ctx sdk.Context, snapshotHeight int64, parameter *types.SlashInputInfo) (*types.SlashExecutionInfo, error) {
```
This height cannot be negative, because it's always obtained through `ctx.BlockHeight()` when stored.

TimmyExogenous commented on proto/exocore/operator/v1/tx.proto:
```diff
@@ -51,6 +51,60 @@ message OperatorOptedUSDValue {
+  ];
+}
+
+// VotingPowerSnapshot is used to record the historical voting power for AVSs
+message VotingPowerSnapshot {
+  // total_voting_power is the total voting power of specified AVS
+  string total_voting_power = 1
+  [
+    (cosmos_proto.scalar) = "cosmos.Dec",
+    (gogoproto.customtype) = "github.com/cosmos/cosmos-sdk/types.Dec",
+    (gogoproto.nullable) = false
+  ];
+  // operator_voting_powers records the active voting power of all operators
+  // for the specified AVS
+  repeated OperatorVotingPower operator_voting_powers = 2;
+  // last_changed_height is used to indicate the height of most recent change when
+  // the voting_power_set is nil, which can help to fall back to the correct epoch height.
+  int64 last_changed_height = 3;
+  // epoch_identifier record the epoch info
+  string epoch_identifier = 4;
+  // epoch_number indicates which epoch this snapshot serve for
+  int64  epoch_number = 5;
+}
```
Thanks, it will be fixed in the next commit.

TimmyExogenous commented on x/operator/keeper/opt.go:
```diff
@@ -145,5 +145,11 @@ func (k *Keeper) OptOut(ctx sdk.Context, operatorAddress sdk.AccAddress, avsAddr
 	}
+
+	// mark opt out in the snapshot helper
+	err = k.SetOptOutFlag(ctx, avsAddr, true)
```
All opting-out transactions will be processed sequentially, so this refactor is unnecessary.

TimmyExogenous commented on x/operator/types/expected_keepers.go:
```diff
@@ -110,26 +111,32 @@ type AVSKeeper interface {
 	// IsAVS returns true if the address is a registered AVS address.
 	IsAVS(ctx sdk.Context, addr string) (bool, error)
 	IsAVSByChainID(ctx sdk.Context, chainID string) (bool, string)
+	GetAVSEpochInfo(ctx sdk.Context, addr string) (*epochstypes.EpochInfo, error)
+	GetAVSUnbondingDuration(ctx sdk.Context, avsAddr string) (uint64, error)
```
`GetAVSEpochInfo` retrives and returns the `EpochInfo`, and the `AvsUnbondingPeriod`  fielld is visible in the `AVSInfo` struct.

TimmyExogenous commented on x/operator/client/cli/query.go:
```diff
@@ -456,3 +460,106 @@
+
+	flags.AddQueryFlagsToCmd(cmd)
+	return cmd
+}
```
outdated, and we don't change the command format temporarily

TimmyExogenous commented on x/operator/keeper/grpc_query.go:
```diff
@@ -254,3 +259,58 @@ func (k *Keeper) QueryOptInfo(goCtx context.Context, req *types.QueryOptInfoRequ
 	return k.GetOptedInfo(ctx, req.OperatorAddr, req.AvsAddress)
 }
+
+func (k *Keeper) QuerySnapshotHelper(goCtx context.Context, req *types.QuerySnapshotHelperRequest) (*types.SnapshotHelper, error) {
+	ctx := sdk.UnwrapSDKContext(goCtx)
+	snapshotHelper, err := k.GetSnapshotHelper(ctx, strings.ToLower(req.Avs))
+	if err != nil {
+		return nil, err
+	}
+	return &snapshotHelper, nil
+}
```
The address check has been added, and returning the error directly is fine.

TimmyExogenous commented on x/operator/client/cli/query.go:
```diff
@@ -456,3 +460,106 @@ func GetOptInfo() *cobra.Command {
 	return cmd
 }
+
+func QuerySnapshotHelper() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "QuerySnapshotHelper <avsAddr>",
+		Short: "Get the voting power snapshot helper for the avs",
+		Long:  "Get the voting power snapshot helper for the avs",
+		Args:  cobra.ExactArgs(1),
+		RunE: func(cmd *cobra.Command, args []string) error {
+			if !common.IsHexAddress(args[0]) {
+				return xerrors.Errorf("invalid avs address,err:%s", types.ErrInvalidAddr)
+			}
+			clientCtx, err := client.GetClientQueryContext(cmd)
+			if err != nil {
+				return err
+			}
+			queryClient := operatortypes.NewQueryClient(clientCtx)
+			req := &operatortypes.QuerySnapshotHelperRequest{
+				Avs: strings.ToLower(args[0]),
+			}
+			res, err := queryClient.QuerySnapshotHelper(context.Background(), req)
+			if err != nil {
+				return err
+			}
+			return clientCtx.PrintProto(res)
+		},
+	}
+
+	flags.AddQueryFlagsToCmd(cmd)
+	return cmd
+}
```
Thanks, it will be changed in the next commit.

TimmyExogenous commented on x/operator/client/cli/query.go:
```diff
@@ -456,3 +460,106 @@
+
+	flags.AddQueryFlagsToCmd(cmd)
+	return cmd
+}
```
Thanks, it will be changed in the next commit.

TimmyExogenous committed:
1. use `ZeroInt` and `LegacyZeroDec` to replace `NewInt(0)` and `LegacyNewDec(0)`
2. refine the command format and description of voting power snapshot

MaxMustermann2 commented on utils/utils.go:
```diff
@@ -210,3 +211,9 @@ func SortByPower(
 	return sortedOperatorAddrs, sortedPubKeys, sortedPowers
 }
+
+func EncodeHeightBytes(height uint64) []byte {
```
Use `sdk.Uint64ToBigEndian`

TimmyExogenous replied:
Thanks, it will be fixed in the next commit.

MaxMustermann2 commented on x/avs/keeper/keeper.go:
```diff
@@ -382,6 +384,19 @@ func (k Keeper) IterateAVSInfo(ctx sdk.Context, fn func(index int64, avsInfo typ
+		return nil, err
+	}
+	avsInfo := avsInfoResp.Info
+	epochInfo, found := k.epochsKeeper.GetEpochInfo(ctx, avsInfo.EpochIdentifier)
```
This is already guarded by x/avs. The check is unnecessary.

TimmyExogenous replied:
thanks, it will be removed in the next commit.

MaxMustermann2 commented on x/dogfood/keeper/abci.go:
```diff
@@ -94,36 +85,30 @@ func (k Keeper) EndBlock(ctx sdk.Context) []abci.ValidatorUpdate {
 	}
 	operators, keys, powers = utils.SortByPower(operators, keys, powers)
 	maxVals := k.GetMaxValidators(ctx)
-	logger.Info("before loop", "maxVals", maxVals, "len(operators)", len(operators))
```
Please restore these `logger` lines including the ones below this one.

TimmyExogenous replied:
Are you referring to the debug print codes from Devin? I removed them when I resolved the conflict after checking with Devin. I can add them back if you think they are necessary. @trestinlsd

MaxMustermann2 replied:
Yes, I had added those logs to Devin's PR for easier debugging.

TimmyExogenous replied:
thanks, I will add them back.

MaxMustermann2 commented on x/dogfood/keeper/keeper.go:
```diff
@@ -86,27 +86,27 @@ func (k Keeper) Hooks() types.DogfoodHooks {
 }
 
-// MarkEpochEnd marks the end of the epoch. It is called within the BeginBlocker to inform
+// MarkUpdateValidatorSetFlag marks the end of the epoch. It is called within the BeginBlocker to inform
 // the module to apply the validator updates at the end of this block.
```
```suggestion
// MarkUpdateValidatorSetFlag marks that the validator set needs to be updated at the end of this block.
// Mostly, these updates occur in response to the epoch ending. In other cases, they are the result of slashing.
```

TimmyExogenous replied:
thanks, it will be fixed in the next commit.

MaxMustermann2 commented on x/operator/keeper/usd_value.go:
```diff
@@ -440,14 +463,14 @@ func (k Keeper) GetOrCalculateOperatorUSDValues(
 	}
 	if optedUSDValues.ActiveUSDValue.IsZero() {
-		return sdkmath.LegacyNewDec(0), err
+		return sdkmath.LegacyZeroDec(), err
```
```suggestion
		return sdkmath.LegacyZeroDec(), nil
```

TimmyExogenous replied:
thanks, it will be fixed in the next commit.

bwhour commented on x/operator/keeper/voting_power_snapshot.go:
```diff
@@ -0,0 +1,285 @@
+	return k.UpdateSnapshotHelper(ctx, avsAddr, opFunc)
+}
+
+func (k *Keeper) SetLastChangedHeight(ctx sdk.Context, avsAddr string, lastChangeHeight int64) error {
```
It is recommended to unify the naming style, using either **Set** or **Update** prefix for all modification functions.

TimmyExogenous replied:
There is already a function named `SetSnapshotHelper`; `UpdateSnapshotHelper` is used solely to update the fields of the snapshot helper.

bwhour reviewed (approved):
it looks good.

TimmyExogenous committed:
Merge remote-tracking branch 'upstream/develop' into develop

TimmyExogenous commented on x/avs/keeper/task_test.go:
```diff
@@ -18,7 +18,7 @@ func (suite *AVSTestSuite) TestTaskInfo() {
 		TaskChallengePeriod: 5000,
 		ThresholdPercentage: 60,
-		TaskTotalPower:      sdk.Dec(sdkmath.NewInt(0)),
+		TaskTotalPower:      sdk.Dec(sdkmath.ZeroInt()),
```
I will use `ZeroDec` here, and all related codes have been replaced.

TimmyExogenous commented on x/operator/keeper/usd_value.go:
```diff
@@ -440,14 +463,14 @@ func (k Keeper) GetOrCalculateOperatorUSDValues(
 	if !k.IsActive(ctx, operator, avsAddr) {
-		return sdkmath.LegacyNewDec(0), nil
+		return sdkmath.LegacyZeroDec(), nil
 	}
 	optedUSDValues, err := k.GetOperatorOptedUSDValue(ctx, avsAddr, operator.String())
 	if err != nil {
 		return sdkmath.LegacyDec{}, err
 	}
 	if optedUSDValues.ActiveUSDValue.IsZero() {
-		return sdkmath.LegacyNewDec(0), err
+		return sdkmath.LegacyZeroDec(), err
```
Thanks, it will be fixed in the next commit.

TimmyExogenous committed:
fix the incorrect comparison in GetImpactfulAVSForOperator
rebase to develop and fix the issues from code comments

TimmyExogenous commented on x/operator/keeper/operator.go:
```diff
@@ -180,21 +180,83 @@ func (k *Keeper) IsActive(ctx sdk.Context, operatorAddr sdk.AccAddress, avsAddr
+
+func (k *Keeper) GetImpactfulAVSForOperator(ctx sdk.Context, operatorAddr string) ([]string, error) {
+	avsList := make([]string, 0)
+	opFunc := func(key []byte, optedInfo *operatortypes.OptedInfo) error {
+		keys, err := assetstype.ParseJoinedStoreKey(key, 2)
+		avsAddr := keys[1]
+		if err != nil {
+			return err
+		}
+		// add AVS currently opting in to the operator's list.
+		if optedInfo.OptedOutHeight == operatortypes.DefaultOptedOutHeight {
+			avsList = append(avsList, avsAddr)
+		} else {
+			// Add AVS that have opted out but are still within the unbonding duration,
+			// and therefore still affect the operator, to the list.
+			// #nosec G115
+			epochNumber, err := k.GetEpochNumberByOptOutHeight(ctx, avsAddr, int64(optedInfo.OptedOutHeight))
+			if err != nil {
+				return err
+			}
+			epochInfo, err := k.avsKeeper.GetAVSEpochInfo(ctx, avsAddr)
+			if err != nil {
+				return err
+			}
+			unbondingDuration, err := k.avsKeeper.GetAVSUnbondingDuration(ctx, avsAddr)
+			if err != nil {
+				return err
+			}
+			// #nosec G115
+			if epochNumber+int64(unbondingDuration) <= epochInfo.CurrentEpoch {
+				avsList = append(avsList, avsAddr)
+			}
+		}
+		return nil
+	}
+	err := k.IterateOptInfo(ctx, false, []byte(operatorAddr), opFunc)
+	if err != nil {
+		return nil, err
 	}
 	return avsList, nil
```
Thanks, the comparison will be fixed in the next commit. Seems like it's unnecessary to check the overflow.

TimmyExogenous commented on x/avs/keeper/keeper.go:
```diff
@@ -382,6 +384,17 @@ func (k Keeper) IterateAVSInfo(ctx sdk.Context, fn func(index int64, avsInfo typ
 	}
 }
 
+func (k Keeper) GetAVSEpochInfo(ctx sdk.Context, addr string) (*epochstypes.EpochInfo, error) {
+	avsInfoResp, err := k.GetAVSInfo(ctx, addr)
+	if err != nil {
+		return nil, err
+	}
+	avsInfo := avsInfoResp.Info
+	// Epoch information must be present because it is checked when setting AVS information.
+	epochInfo, _ := k.epochsKeeper.GetEpochInfo(ctx, avsInfo.EpochIdentifier)
+	return &epochInfo, nil
+}
```
It’s unnecessary to check the AVS address because this function is called internally and will return an error if an invalid address is input. Regarding the `found` check, it is already guarded when setting, so we don't need to check it here.

TimmyExogenous commented on x/operator/keeper/operator.go:
```diff
@@ -180,21 +180,83 @@ func (k *Keeper) IsActive(ctx sdk.Context, operatorAddr sdk.AccAddress, avsAddr
+
+func (k *Keeper) GetImpactfulAVSForOperator(ctx sdk.Context, operatorAddr string) ([]string, error) {
+	avsList := make([]string, 0)
+	opFunc := func(key []byte, optedInfo *operatortypes.OptedInfo) error {
+		keys, err := assetstype.ParseJoinedStoreKey(key, 2)
+		avsAddr := keys[1]
+		if err != nil {
+			return err
+		}
+		// add AVS currently opting in to the operator's list.
+		if optedInfo.OptedOutHeight == operatortypes.DefaultOptedOutHeight {
+			avsList = append(avsList, avsAddr)
+		} else {
+			// Add AVS that have opted out but are still within the unbonding duration,
+			// and therefore still affect the operator, to the list.
+			// #nosec G115
+			epochNumber, err := k.GetEpochNumberByOptOutHeight(ctx, avsAddr, int64(optedInfo.OptedOutHeight))
+			if err != nil {
+				return err
+			}
+			epochInfo, err := k.avsKeeper.GetAVSEpochInfo(ctx, avsAddr)
+			if err != nil {
+				return err
+			}
+			unbondingDuration, err := k.avsKeeper.GetAVSUnbondingDuration(ctx, avsAddr)
+			if err != nil {
+				return err
+			}
+			// #nosec G115
+			if epochNumber >= epochInfo.CurrentEpoch-int64(unbondingDuration) {
+				avsList = append(avsList, avsAddr)
+			}
+		}
+		return nil
+	}
+	err := k.IterateOptInfo(ctx, false, []byte(operatorAddr), opFunc)
+	if err != nil {
+		return nil, err
 	}
 	return avsList, nil
```
It should include the equal case here.

TimmyExogenous commented on x/operator/keeper/grpc_query.go:
```diff
@@ -338,3 +344,58 @@ func (k *Keeper) Validator(c context.Context, req *types.QueryValidatorRequest)
+	return &snapshotHelper, nil
+}
+
+func (k *Keeper) QuerySpecifiedSnapshot(goCtx context.Context, req *types.QuerySpecifiedSnapshotRequest) (*types.VotingPowerSnapshotKeyHeight, error) {
+	ctx := sdk.UnwrapSDKContext(goCtx)
+	findHeight, snapshot, err := k.LoadVotingPowerSnapshot(ctx, strings.ToLower(req.Avs), req.Height)
+	if err != nil {
+		return nil, err
+	}
+	return &types.VotingPowerSnapshotKeyHeight{
+		SnapshotKeyHeight: findHeight,
+		Snapshot:          snapshot,
+	}, nil
+}
```
There's no snapshot found when the input height is negative, so it seems this check isn't necessary.

cloud8little commented:
Test scenario: 
- make undelegation at the start of an epoch.
- shut down the operator node
- wait for the operator to be slashed.
- check the slash undelegation information in the operator slash info.
- after an epoch, check the staker withdrawable amount, it should be slashed.

Test Result: 
![photo_2024-11-20_12-10-05](https://github.com/user-attachments/assets/69fdc437-d4a8-45d0-b008-a77db01f6fd7)

![photo_2024-11-20_14-48-38](https://github.com/user-attachments/assets/28eb3643-6121-4f94-b841-674245a8e67c)
